# Single Path One-Shot Neural Architecture Search using Random Search

# Train found architecture from scratch

## Setup

### Imports

In [1]:
from typing import Tuple

import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from thop import profile
from torch.utils.data import DataLoader
from tqdm import tqdm

from cifar10 import get_train_transform, get_val_transform
from supernet import supernet18

### Make everything a bit faster

In [2]:
torch.backends.cudnn.benchmark = True
torch.backends.cudnn.deterministic = False

### Build datasets and dataloaders for CIFAR-10

In [3]:
# Change this value if needed.
batch_size = 512

In [4]:
train_transform = get_train_transform()
val_transform = get_val_transform()

train_set = torchvision.datasets.CIFAR10(
    root='./data',
    train=True,
    download=True,
    transform=train_transform,
)
test_set = torchvision.datasets.CIFAR10(
    root='./data',
    train=False,
    download=True,
    transform=val_transform,
)

train_dataloader = DataLoader(
    train_set,
    batch_size=batch_size,
    shuffle=True,
    num_workers=4,
    drop_last=True,
)
test_dataloader = DataLoader(
    test_set,
    batch_size=batch_size,
    shuffle=False,
    num_workers=4,
    drop_last=False,
)

Files already downloaded and verified
Files already downloaded and verified


## Create supernet

In [5]:
# Select suitable device.
# You should probably use either cuda (NVidia GPU) or mps (Apple) backend.
device = torch.device('cuda:0')

In [6]:
channel_multipliers = (0.25, 0.5, 0.75, 1.0, 1.25, 1.5)
model = supernet18(num_classes=10, zero_init_residual=True, channel_multipliers=channel_multipliers)
model.to(device=device)

Supernet(
  (conv1): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
  (maxpool): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (layer1): Sequential(
    (0): SearchBlock(
      (ops): ModuleList(
        (0): BasicBlock(
          (conv1): Conv2d(64, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
          (bn1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
          (relu): ReLU(inplace=True)
          (conv2): Conv2d(16, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
          (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        )
        (1): BasicBlock(
          (conv1): Conv2d(64, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
          (bn1): BatchNorm2d(32, eps=1e-05,

## Sample found architecture from the supernet

In [7]:
# This is a necessary step!
best_architecture = [4, 1, 4, 5, 1, 0, 1, 0]
model.sample(best_architecture)

### Compute the number of MACs and parameters for the sampled architecture

In [8]:
macs, params = profile(model, inputs=(torch.zeros(1, 3, 32, 32, device=device),))
print(f'Number of macs: {macs / 1e6:.2f}M, number of parameters: {params / 1e6:.2f}M')

[INFO] Register count_convNd() for <class 'torch.nn.modules.conv.Conv2d'>.
[INFO] Register count_normalization() for <class 'torch.nn.modules.batchnorm.BatchNorm2d'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.activation.ReLU'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.pooling.MaxPool2d'>.
[INFO] Register zero_ops() for <class 'torch.nn.modules.container.Sequential'>.
[INFO] Register count_adap_avgpool() for <class 'torch.nn.modules.pooling.AdaptiveAvgPool2d'>.
[INFO] Register count_linear() for <class 'torch.nn.modules.linear.Linear'>.
Number of macs: 28.66M, number of parameters: 4.73M


## Train found architecture

### Define loss function

In [9]:
criterion = nn.CrossEntropyLoss()

### Select hyperparameters

In [10]:
lr = 0.25
weight_decay = 5e-4
momentum = 0.9
n_epochs = 20  # Train for the same number of epochs as baseline for fair comparison.

### Build optimizer and scheduler

In [11]:
optimizer = optim.SGD(model.parameters(), lr=lr, weight_decay=weight_decay, momentum=momentum)
scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=len(train_dataloader) * n_epochs)

### Define training and evaluation functions

In [12]:
def train_one_epoch(
        model: nn.Module,
        criterion: nn.Module,
        dataloader: DataLoader,
        optimizer: optim.Optimizer,
        scheduler,
        device: torch.device,
        epoch: int,
) -> Tuple[float, float]:
    model.train()

    total_loss = 0.0
    total_correct = 0.0
    total_samples = 0

    wrapped_dataloader = tqdm(enumerate(dataloader), total=len(dataloader))
    for i, (inputs, labels) in wrapped_dataloader:
        inputs = inputs.to(device=device)
        labels = labels.to(device=device)

        optimizer.zero_grad()

        logits = model(inputs)
        loss = criterion(logits, labels)
        loss.backward()
        optimizer.step()
        scheduler.step()

        with torch.no_grad():
            _, predicted_labels = torch.max(logits, 1)
            total_loss += loss.item()
            total_correct += (predicted_labels == labels).sum().item()
            total_samples += labels.shape[0]

        wrapped_dataloader.set_description(
            f'(train) Epoch={epoch}, lr={scheduler.get_last_lr()[0]:.4f} loss={total_loss / (i + 1):.3f}'
        )

    return total_loss / len(dataloader), total_correct / total_samples


@torch.no_grad()
def validate_one_epoch(
        model: nn.Module,
        criterion: nn.Module,
        dataloader: DataLoader,
        device: torch.device,
        epoch: int,
) -> Tuple[float, float]:
    model.eval()

    total_loss = 0.0
    total_correct = 0.0
    total_samples = 0

    wrapped_dataloader = tqdm(enumerate(dataloader), total=len(dataloader))
    for i, (inputs, labels) in wrapped_dataloader:
        inputs = inputs.to(device=device)
        labels = labels.to(device=device)

        logits = model(inputs)
        loss = criterion(logits, labels)
        _, predicted_labels = torch.max(logits, 1)
        total_loss += loss.item()
        total_correct += (predicted_labels == labels).sum().item()
        total_samples += labels.shape[0]

        wrapped_dataloader.set_description(f'(val) Epoch={epoch}, loss={total_loss / (i + 1):.3f}')

    return total_loss / len(dataloader), total_correct / total_samples

### Run training

In [13]:
for epoch in range(n_epochs):
    print(f'Epoch: {epoch}')
    loss, accuracy = train_one_epoch(model, criterion, train_dataloader, optimizer, scheduler, device, epoch)
    print(f'train_loss={loss:.4f}, train_accuracy={accuracy:.3%}')
    loss, accuracy = validate_one_epoch(model, criterion, test_dataloader, device, epoch)
    print(f'test_loss={loss:.4f}, test_accuracy={accuracy:.3%}')

Epoch: 0


(train) Epoch=0, lr=0.2485 loss=1.947: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:03<00:00, 30.24it/s]

train_loss=1.9470, train_accuracy=28.810%



(val) Epoch=0, loss=1.530: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 41.27it/s]

test_loss=1.5305, test_accuracy=43.730%
Epoch: 1



(train) Epoch=1, lr=0.2439 loss=1.425: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.80it/s]

train_loss=1.4252, train_accuracy=48.206%



(val) Epoch=1, loss=1.266: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 54.85it/s]

test_loss=1.2657, test_accuracy=54.980%
Epoch: 2



(train) Epoch=2, lr=0.2364 loss=1.234: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.86it/s]

train_loss=1.2340, train_accuracy=55.841%



(val) Epoch=2, loss=1.165: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 57.09it/s]

test_loss=1.1645, test_accuracy=59.190%
Epoch: 3



(train) Epoch=3, lr=0.2261 loss=1.082: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.42it/s]

train_loss=1.0818, train_accuracy=61.268%



(val) Epoch=3, loss=1.046: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 55.67it/s]

test_loss=1.0459, test_accuracy=63.040%
Epoch: 4



(train) Epoch=4, lr=0.2134 loss=0.990: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.32it/s]


train_loss=0.9901, train_accuracy=64.771%


(val) Epoch=4, loss=0.932: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 56.75it/s]

test_loss=0.9324, test_accuracy=67.430%
Epoch: 5



(train) Epoch=5, lr=0.1985 loss=0.916: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.40it/s]

train_loss=0.9165, train_accuracy=67.721%



(val) Epoch=5, loss=0.902: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 57.05it/s]

test_loss=0.9015, test_accuracy=68.020%
Epoch: 6



(train) Epoch=6, lr=0.1817 loss=0.857: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.04it/s]

train_loss=0.8570, train_accuracy=69.773%



(val) Epoch=6, loss=0.876: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 54.40it/s]

test_loss=0.8758, test_accuracy=69.490%
Epoch: 7



(train) Epoch=7, lr=0.1636 loss=0.810: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 33.55it/s]

train_loss=0.8095, train_accuracy=71.515%



(val) Epoch=7, loss=0.861: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 59.27it/s]


test_loss=0.8615, test_accuracy=70.220%
Epoch: 8


(train) Epoch=8, lr=0.1446 loss=0.760: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.52it/s]

train_loss=0.7600, train_accuracy=73.548%



(val) Epoch=8, loss=0.804: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 53.42it/s]

test_loss=0.8039, test_accuracy=72.270%
Epoch: 9



(train) Epoch=9, lr=0.1250 loss=0.714: 100%|███████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 36.03it/s]

train_loss=0.7142, train_accuracy=74.934%



(val) Epoch=9, loss=0.765: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 56.53it/s]

test_loss=0.7650, test_accuracy=73.470%
Epoch: 10



(train) Epoch=10, lr=0.1054 loss=0.678: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 34.85it/s]

train_loss=0.6782, train_accuracy=76.248%



(val) Epoch=10, loss=0.687: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 55.05it/s]

test_loss=0.6874, test_accuracy=76.330%
Epoch: 11



(train) Epoch=11, lr=0.0864 loss=0.642: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.90it/s]

train_loss=0.6416, train_accuracy=77.515%



(val) Epoch=11, loss=0.678: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 57.90it/s]

test_loss=0.6781, test_accuracy=77.080%
Epoch: 12



(train) Epoch=12, lr=0.0683 loss=0.602: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.33it/s]

train_loss=0.6018, train_accuracy=79.003%



(val) Epoch=12, loss=0.619: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 54.45it/s]

test_loss=0.6186, test_accuracy=79.090%
Epoch: 13



(train) Epoch=13, lr=0.0515 loss=0.557: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 34.85it/s]

train_loss=0.5565, train_accuracy=80.533%



(val) Epoch=13, loss=0.636: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 52.65it/s]

test_loss=0.6358, test_accuracy=78.630%
Epoch: 14



(train) Epoch=14, lr=0.0366 loss=0.520: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 34.85it/s]

train_loss=0.5201, train_accuracy=81.741%



(val) Epoch=14, loss=0.610: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 59.24it/s]

test_loss=0.6096, test_accuracy=79.550%
Epoch: 15



(train) Epoch=15, lr=0.0239 loss=0.483: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.27it/s]

train_loss=0.4828, train_accuracy=82.949%



(val) Epoch=15, loss=0.540: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 53.51it/s]

test_loss=0.5401, test_accuracy=81.780%
Epoch: 16



(train) Epoch=16, lr=0.0136 loss=0.448: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.56it/s]

train_loss=0.4483, train_accuracy=84.409%



(val) Epoch=16, loss=0.530: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 54.68it/s]

test_loss=0.5301, test_accuracy=82.020%
Epoch: 17



(train) Epoch=17, lr=0.0061 loss=0.427: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.17it/s]

train_loss=0.4271, train_accuracy=84.963%



(val) Epoch=17, loss=0.507: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 57.61it/s]

test_loss=0.5068, test_accuracy=82.750%
Epoch: 18



(train) Epoch=18, lr=0.0015 loss=0.403: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 34.45it/s]

train_loss=0.4026, train_accuracy=86.042%



(val) Epoch=18, loss=0.500: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 57.98it/s]


test_loss=0.4999, test_accuracy=83.030%
Epoch: 19


(train) Epoch=19, lr=0.0000 loss=0.390: 100%|██████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:02<00:00, 35.09it/s]

train_loss=0.3903, train_accuracy=86.314%



(val) Epoch=19, loss=0.500: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 54.20it/s]

test_loss=0.4998, test_accuracy=83.090%


### Save trained model weights

In [14]:
torch.save(model.state_dict(), 'advanced_arch_41451010.pth')

# Final analysis

### Well, now we're matching baseline accuracy!

We were able to find a faster architecture (37.1M -> 28.7M MACs) with the same accuracy (83.1%).

# In notebooks 10-12 you will find how batch norm tuning can help with NAS